src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java

```diff
@@ -71,46 +71,55 @@ static String stringOrNull(RequestParameter requestParameter) {
     String tenant = stringOrNull(params.headerParameter(XOkapiHeaders.TENANT));
+    String counterReportId = stringOrNull(params.queryParameter("counterReportId"));
 
     TenantPgPool pool = TenantPgPool.pool(vertx, tenant);
     return pool.getConnection()
-        .compose(sqlConnection ->
-            sqlConnection.prepare("SELECT * FROM " + titleEntriesTable(pool))
-                .<Void>compose(pq ->
-                    sqlConnection.begin().compose(tx -> {
-                      ctx.response().setChunked(true);
-                      ctx.response().putHeader("Content-Type", "application/json");
-                      ctx.response().write("{ \"titles\" : [");
-                      AtomicInteger offset = new AtomicInteger();
-                      RowStream<Row> stream = pq.createStream(50);
-                      stream.handler(row -> {
-                        if (offset.incrementAndGet() > 1) {
-                          ctx.response().write(",");
-                        }
-                        JsonObject response = new JsonObject()
-                            .put("id", row.getUUID(0))
-                            .put("counterReportTitle", row.getString(1));
-                        String titleName = row.getString(3);
-                        if (titleName != null) {
-                          response.put("kbTitleName", titleName)
-                              .put("kbTitleId", row.getUUID(4));
-                        }
-                        Boolean kbManualMatch = row.getBoolean(5);
-                        if (kbManualMatch != null) {
-                          response.put("kbManualMatch", kbManualMatch);
-                        }
-                        ctx.response().write(response.encode());
-                      });
-                      stream.endHandler(end -> {
-                        ctx.response().write("] }");
-                        ctx.response().end();
-                        tx.commit();
-                      });
-                      return Future.succeededFuture();
-                    })
-                )
-                .eventually(x -> sqlConnection.close())
-        );
+        .compose(sqlConnection -> {
+          String qry = "SELECT DISTINCT ON (" + titleEntriesTable(pool) + ".id) * FROM "
+              + titleEntriesTable(pool);
+          if (counterReportId != null) {
+            qry = qry + " INNER JOIN " + titleDataTable(pool)
+                + " ON reportTitleId = " + titleEntriesTable(pool) + ".id"
+                + " WHERE counterReportId = '" + UUID.fromString(counterReportId) + "'";
+          }
+          return sqlConnection.prepare(qry)
+              .<Void>compose(pq ->
+                  sqlConnection.begin().compose(tx -> {
+                    ctx.response().setChunked(true);
+                    ctx.response().putHeader("Content-Type", "application/json");
+                    ctx.response().write("{ \"titles\" : [");
```

Comment:
I've not looked at this code before. It comes as quite the surprise that you assemble the JSON response string by hand, instead of building an object and serialising it.

Reply:
Because that could use a lot of memory... Using cursors for the database.. Hence written in chunks (and so can not have one huge JsonObject).

Reply:
Solid reasoning!